Camunda Exporter is not creating indices for templates, when templates already exist

## Description

We ran into this during implementing the new IT foundation #26896 as we detected that indices are not cleaned up correctly, and later some indices are not re-created.

It turned out that if the templates were not cleaned up as well, the corresponding indices haven't been recreated (even if they were missing). Causing endless loops of waiting for all indices.

In an production environment this might be less likely, but in general I think it makes sense to de-couple this logic to always create the index (and log the failure if it already exists).

Happy to discuss this further @EuroLew

@ChrisKujawa @EuroLew there is an error on our DEV, looks very similar. Can you please verify? Otherwise, I'll create a new issue:

Environment: `8.7.0-alpha1`

[GCP logs](https://console.cloud.google.com/logs/query;cursorTimestamp=2025-03-04T16:37:33.840953269Z;endTime=2025-03-04T17:08:03.840Z;pinnedLogId=2025-03-04T16:37:33.840953269Z%2Fswxihgbn489vz62a;query=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.location%3D%22europe-west2%22%0Aresource.labels.project_id%3D%22camunda-saas-prod-lhr%22%0Aresource.labels.pod_name%3D%22zeebe-1%22%0Aresource.labels.cluster_name%3D%22worker-6%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.namespace_name%3D%227953e69a-fb25-46dc-a1e7-01dc89fad486-zeebe%22%0Atimestamp%3D%222025-03-04T16:37:33.840953269Z%22%0AinsertId%3D%22swxihgbn489vz62a%22;startTime=2025-03-04T16:08:03.840Z?project=camunda-saas-prod-lhr)

<details><summary>Full Stacktrace</summary>
 <p>

```
io.camunda.exporter.exceptions.ElasticsearchExporterException: Template [operate-list-view-8.3.0_template] was NOT created
	at io.camunda.exporter.schema.elasticsearch.ElasticsearchEngineClient.createIndexTemplate(ElasticsearchEngineClient.java:90) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.exporter.schema.SchemaManager.lambda$initialiseIndexTemplates$3(SchemaManager.java:96) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(Unknown Source) ~[?:?]
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(Unknown Source) ~[?:?]
	at java.base/java.util.Iterator.forEachRemaining(Unknown Source) ~[?:?]
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Unknown Source) ~[?:?]
	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(Unknown Source) ~[?:?]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(Unknown Source) ~[?:?]
	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
	at java.base/java.util.stream.ReferencePipeline.forEach(Unknown Source) ~[?:?]
	at io.camunda.exporter.schema.SchemaManager.initialiseIndexTemplates(SchemaManager.java:94) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.exporter.schema.SchemaManager.initialiseResources(SchemaManager.java:63) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.exporter.schema.SchemaManager.startup(SchemaManager.java:49) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.exporter.CamundaExporter.open(CamundaExporter.java:87) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.broker.exporter.stream.ExporterContainer.lambda$openExporter$0(ExporterContainer.java:118) ~[zeebe-broker-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.util.jar.ThreadContextUtil.runCheckedWithClassLoader(ThreadContextUtil.java:59) ~[zeebe-util-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.util.jar.ThreadContextUtil.runWithClassLoader(ThreadContextUtil.java:35) ~[zeebe-util-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.broker.exporter.stream.ExporterContainer.openExporter(ExporterContainer.java:117) ~[zeebe-broker-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.broker.exporter.stream.ExporterDirector.lambda$startActiveExportingMode$16(ExporterDirector.java:530) ~[zeebe-broker-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.scheduler.retry.BackOffRetryStrategy.run(BackOffRetryStrategy.java:51) ~[zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:85) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:42) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:122) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
Caused by: co.elastic.clients.elasticsearch._types.ElasticsearchException: [es/indices.put_index_template] failed: [illegal_argument_exception] index template [operate-list-view-8.3.0_template] already exists
	at co.elastic.clients.transport.ElasticsearchTransportBase.getApiResponse(ElasticsearchTransportBase.java:345) ~[elasticsearch-java-8.13.4.jar:?]
	at co.elastic.clients.transport.ElasticsearchTransportBase.performRequest(ElasticsearchTransportBase.java:147) ~[elasticsearch-java-8.13.4.jar:?]
	at co.elastic.clients.elasticsearch.indices.ElasticsearchIndicesClient.putIndexTemplate(ElasticsearchIndicesClient.java:1560) ~[elasticsearch-java-8.13.4.jar:?]
	at io.camunda.exporter.schema.elasticsearch.ElasticsearchEngineClient.createIndexTemplate(ElasticsearchEngineClient.java:84) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
	... 27 more

```

</p>
</details>

> @ChrisKujawa @EuroLew there is an error on one of SaaS Prod Clusters (but for internal usage), looks very similar. Can you please verify? Otherwise, I'll create a new issue.
> 
> Environment: `8.7.0-alpha1`
> 
> [GCP logs](https://console.cloud.google.com/logs/query;cursorTimestamp=2025-03-04T16:37:33.840953269Z;endTime=2025-03-04T17:08:03.840Z;pinnedLogId=2025-03-04T16:37:33.840953269Z%2Fswxihgbn489vz62a;query=logName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.location%3D%22europe-west2%22%0Aresource.labels.project_id%3D%22camunda-saas-prod-lhr%22%0Aresource.labels.pod_name%3D%22zeebe-1%22%0Aresource.labels.cluster_name%3D%22worker-6%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.namespace_name%3D%227953e69a-fb25-46dc-a1e7-01dc89fad486-zeebe%22%0Atimestamp%3D%222025-03-04T16:37:33.840953269Z%22%0AinsertId%3D%22swxihgbn489vz62a%22;startTime=2025-03-04T16:08:03.840Z?project=camunda-saas-prod-lhr)
> 
> <details><summary>Full Stacktrace</summary>
>  <p>
> 
> ```
> io.camunda.exporter.exceptions.ElasticsearchExporterException: Template [operate-list-view-8.3.0_template] was NOT created
> 	at io.camunda.exporter.schema.elasticsearch.ElasticsearchEngineClient.createIndexTemplate(ElasticsearchEngineClient.java:90) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.exporter.schema.SchemaManager.lambda$initialiseIndexTemplates$3(SchemaManager.java:96) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(Unknown Source) ~[?:?]
> 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(Unknown Source) ~[?:?]
> 	at java.base/java.util.Iterator.forEachRemaining(Unknown Source) ~[?:?]
> 	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Unknown Source) ~[?:?]
> 	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
> 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
> 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(Unknown Source) ~[?:?]
> 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(Unknown Source) ~[?:?]
> 	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
> 	at java.base/java.util.stream.ReferencePipeline.forEach(Unknown Source) ~[?:?]
> 	at io.camunda.exporter.schema.SchemaManager.initialiseIndexTemplates(SchemaManager.java:94) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.exporter.schema.SchemaManager.initialiseResources(SchemaManager.java:63) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.exporter.schema.SchemaManager.startup(SchemaManager.java:49) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.exporter.CamundaExporter.open(CamundaExporter.java:87) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.broker.exporter.stream.ExporterContainer.lambda$openExporter$0(ExporterContainer.java:118) ~[zeebe-broker-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.util.jar.ThreadContextUtil.runCheckedWithClassLoader(ThreadContextUtil.java:59) ~[zeebe-util-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.util.jar.ThreadContextUtil.runWithClassLoader(ThreadContextUtil.java:35) ~[zeebe-util-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.broker.exporter.stream.ExporterContainer.openExporter(ExporterContainer.java:117) ~[zeebe-broker-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.broker.exporter.stream.ExporterDirector.lambda$startActiveExportingMode$16(ExporterDirector.java:530) ~[zeebe-broker-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.scheduler.retry.BackOffRetryStrategy.run(BackOffRetryStrategy.java:51) ~[zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:85) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:42) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:122) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227) [zeebe-scheduler-8.7.0-alpha1.jar:8.7.0-alpha1]
> Caused by: co.elastic.clients.elasticsearch._types.ElasticsearchException: [es/indices.put_index_template] failed: [illegal_argument_exception] index template [operate-list-view-8.3.0_template] already exists
> 	at co.elastic.clients.transport.ElasticsearchTransportBase.getApiResponse(ElasticsearchTransportBase.java:345) ~[elasticsearch-java-8.13.4.jar:?]
> 	at co.elastic.clients.transport.ElasticsearchTransportBase.performRequest(ElasticsearchTransportBase.java:147) ~[elasticsearch-java-8.13.4.jar:?]
> 	at co.elastic.clients.elasticsearch.indices.ElasticsearchIndicesClient.putIndexTemplate(ElasticsearchIndicesClient.java:1560) ~[elasticsearch-java-8.13.4.jar:?]
> 	at io.camunda.exporter.schema.elasticsearch.ElasticsearchEngineClient.createIndexTemplate(ElasticsearchEngineClient.java:84) ~[camunda-exporter-8.7.0-alpha1.jar:8.7.0-alpha1]
> 	... 27 more
> 
> ```
> 
> </p>
> </details>
> 

I dont think it makes sense to do this for such old alpha version

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/SchemaManager.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/SchemaManager.java
@@ -11,6 +11,8 @@ import com.fasterxml.jackson.databind.ObjectMapper;
 import io.camunda.exporter.config.ExporterConfiguration;
 import io.camunda.exporter.config.ExporterConfiguration.IndexSettings;
 import io.camunda.exporter.config.ExporterConfiguration.RetentionConfiguration;
+import io.camunda.exporter.exceptions.ElasticsearchExporterException;
+import io.camunda.exporter.exceptions.OpensearchExporterException;
 import io.camunda.webapps.schema.descriptors.IndexDescriptor;
 import io.camunda.webapps.schema.descriptors.IndexTemplateDescriptor;
 import java.util.Collection;
@@ -92,10 +94,7 @@ public class SchemaManager {
             .getMappings(config.getIndex().getPrefix() + "*", MappingSource.INDEX_TEMPLATE)
             .keySet();
 
-    final var existingIndexNames =
-        allIndexNames().isBlank()
-            ? Set.of()
-            : searchEngineClient.getMappings(allIndexNames(), MappingSource.INDEX).keySet();
+    final var existingIndexNames = existingIndexNames();
 
     indexTemplateDescriptors.stream()
         .filter(desc -> existingTemplateNames.contains(desc.getTemplateName()))
@@ -133,9 +132,7 @@ public class SchemaManager {
       return;
     }
 
-    final var existingIndexNames =
-        Collections.synchronizedSet(
-            searchEngineClient.getMappings(allIndexNames(), MappingSource.INDEX).keySet());
+    final var existingIndexNames = Collections.synchronizedSet(existingIndexNames());
 
     LOG.info(
         "Found '{}' existing indices. Create missing index templates based on '{}' descriptors.",
@@ -179,32 +176,45 @@ public class SchemaManager {
         indexTemplateDescriptors.size());
     final var futures =
         indexTemplateDescriptors.stream()
-            .filter(descriptor -> !existingTemplateNames.contains(descriptor.getTemplateName()))
             .map(
                 descriptor ->
                     // run creation of indices async as virtual thread
                     CompletableFuture.runAsync(
-                        () -> {
-                          LOG.info(
-                              "Create missing index template '{}'", descriptor.getTemplateName());
-                          searchEngineClient.createIndexTemplate(
-                              descriptor,
-                              getIndexSettingsFromConfig(descriptor.getIndexName()),
-                              true);
-                          LOG.info(
-                              "Create missing index '{}', for template '{}'",
-                              descriptor.getFullQualifiedName(),
-                              descriptor.getTemplateName());
-                          searchEngineClient.createIndex(
-                              descriptor, getIndexSettingsFromConfig(descriptor.getIndexName()));
-                        },
-                        virtualThreadExecutor))
+                        () -> createIndexTemplate(descriptor), virtualThreadExecutor))
             .toArray(CompletableFuture[]::new);
 
     // We need to wait for the completion, to make sure all indices and templates have been created
     // successfully
     // Doing this in parallel is still speeding up the bootstrap time
     joinOnFutures(futures);
+  }
+
+  /**
+   * Create an index template and it's matching index.
+   *
+   * @param descriptor a description of the index template to create
+   */
+  private void createIndexTemplate(final IndexTemplateDescriptor descriptor) {
+    try {
+      searchEngineClient.createIndexTemplate(
+          descriptor, getIndexSettingsFromConfig(descriptor.getIndexName()), true);
+      LOG.info(
+          "Index template '{}', has been created / already exists", descriptor.getTemplateName());
+
+      searchEngineClient.createIndex(
+          descriptor, getIndexSettingsFromConfig(descriptor.getIndexName()));
+
+      LOG.info(
+          "Index '{}', for template '{}' has been created / already exists",
+          descriptor.getFullQualifiedName(),
+          descriptor.getTemplateName());
+
+    } catch (final ElasticsearchExporterException | OpensearchExporterException e) {
+      final var errMsg =
+          String.format("Index template '%s' could not be created", descriptor.getTemplateName());
+      LOG.error(errMsg, e);
+      throw new IllegalStateException(errMsg, e);
+    }
   }
 
   /**
@@ -325,6 +335,12 @@ public class SchemaManager {
         indexTemplateDescriptors.stream()
             .map(IndexDescriptor.class::cast)
             .collect(Collectors.toSet()));
+  }
+
+  private Set<String> existingIndexNames() {
+    return allIndexNames().isBlank()
+        ? Set.of()
+        : searchEngineClient.getMappings(allIndexNames(), MappingSource.INDEX).keySet();
   }
 
   private String allIndexNames() {

--- a/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/SchemaManager.java
+++ b/zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/schema/SchemaManager.java
@@ -164,16 +164,8 @@ public class SchemaManager {
       return;
     }
 
-    final var existingTemplateNames =
-        Collections.synchronizedSet(
-            searchEngineClient
-                .getMappings(config.getIndex().getPrefix() + "*", MappingSource.INDEX_TEMPLATE)
-                .keySet());
-
     LOG.info(
-        "Found '{}' existing index templates. Create missing index templates based on '{}' descriptors.",
-        existingTemplateNames.size(),
-        indexTemplateDescriptors.size());
+        "Creating index templates based on '{}' descriptors.", indexTemplateDescriptors.size());
     final var futures =
         indexTemplateDescriptors.stream()
             .map(

--- a/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/schema/SchemaManagerIT.java
+++ b/zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/schema/SchemaManagerIT.java
@@ -22,6 +22,7 @@ import io.camunda.exporter.ExporterMetadata;
 import io.camunda.exporter.cache.ExporterEntityCacheProvider;
 import io.camunda.exporter.config.ConnectionTypes;
 import io.camunda.exporter.config.ExporterConfiguration;
+import io.camunda.exporter.config.ExporterConfiguration.IndexSettings;
 import io.camunda.exporter.schema.elasticsearch.ElasticsearchEngineClient;
 import io.camunda.exporter.schema.opensearch.OpensearchEngineClient;
 import io.camunda.exporter.utils.CamundaExporterITInvocationProvider;
@@ -40,6 +41,7 @@ import java.util.Collections;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
+import org.awaitility.Awaitility;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.TestTemplate;
 import org.junit.jupiter.api.condition.DisabledIfSystemProperty;
@@ -73,8 +75,7 @@ public class SchemaManagerIT {
         SchemaTestUtil.mockIndex(
             CONFIG_PREFIX + "-qualified_name", "alias", "index_name", "/mappings.json");
 
-    when(indexTemplate.getFullQualifiedName())
-        .thenReturn(CONFIG_PREFIX + "-template-index-qualified-name");
+    when(indexTemplate.getFullQualifiedName()).thenReturn(CONFIG_PREFIX + "-qualified_name");
   }
 
   private SearchEngineClient searchEngineClientFromConfig(final ExporterConfiguration config) {
@@ -585,6 +586,39 @@ public class SchemaManagerIT {
 
     assertThat(updatedTemplate.at(replicaSettingPath).asInt()).isEqualTo(5);
     assertThat(updatedTemplate.at(shardsSettingPath).asInt()).isEqualTo(5);
+  }
+
+  @TestTemplate
+  void shouldCreateCorrespondingIndexIfIndexTemplateAlreadyExists(
+      final ExporterConfiguration config, final SearchClientAdapter searchClientAdapter)
+      throws IOException {
+    // given
+    final var searchEngineClient = searchEngineClientFromConfig(config);
+
+    searchEngineClient.createIndexTemplate(indexTemplate, new IndexSettings(), true);
+
+    searchClientAdapter.refresh();
+
+    // when
+    final var schemaManager =
+        new SchemaManager(
+            searchEngineClientFromConfig(config),
+            Set.of(),
+            Set.of(indexTemplate),
+            config,
+            objectMapper);
+
+    schemaManager.startup();
+
+    // then
+    Awaitility.await()
+        .untilAsserted(
+            () ->
+                assertThatNoException()
+                    .isThrownBy(
+                        () ->
+                            searchClientAdapter.getIndexAsNode(
+                                indexTemplate.getFullQualifiedName())));
   }
 
   @RegressionTestTemplate("https://github.com/camunda/camunda/issues/26056")

